docs/design-proposals/2021-11-24-graceful-reschedule-tikv-pod.md

```diff
@@ -0,0 +1,107 @@
+# gracefully reschedule tikv pod
+
+TiDB-Operator issue: [#4215](https://github.com/pingcap/tidb-operator/issues/4215)
+
+## Summary
+
+Support gracefully reschedule a tikv pod. The implemtation is necessarrily the combination of three operations:
+
+1. Add a PD evict-leader-scheduler to transfer the leader.
+2. When leader count drop down to zero, delete the pod to let it re-create and reschedule.
+3. Remove the evict-leader-scheduler.
+
+## Motivation
+
+### Goals
+
+- Provide a way for user to gracefully reschedule a tikv pod.
+- The pod must be gracefully stop, to be specify, evict leaders before delete the pod and must work nomal when recreate the pod.
+
+### Non-Goals
+
+- Provider a way to make tikv graceful stop when just recive a `SIGTERM` signal.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+suppose `tikv-0` is running at node `node0`,  there are many free resource at other nodes, I would like to make tikv-0 run at other node and shutdown `node0` to reduce the cost. I may apply the flowing operations:
+
+1. run `kubectl cordon node0` to mark `node0`as unschedulable.
+2. delete pod `tikv-0` to let it re-create and reschedule to other nodes.
+3. draine `node-0` in the nomal way as [safely drain a node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) and shotdown `node0`
+
+### Risks and Mitigations
+
+- PD/TiKV fails to evict all leader and we may just keep checking it.
+
+## Design Details
+
+Support user to add an annotation to CRD `TidbCluster` to trigger an gracefule reschedule.
```

Comment:
Why not update pod annotation to reschedule it?

Reply:
> then we may lose the update(leave an evict-leader scheduler at pd)
While the pod has the annotation, add evict-leader scheduler at pd. If the pod is deleted or updated(annotation is removed), just remove scheduler at pd. It's level triggered and cancelable so scheduler will not be lefted at pd.

Reply:
You can not always catch the `delete`(like at the time pod deleted operator not running normal), you need to check time from time.

another possible issue is we may also delete evict-leader scheduler not added by us.

Reply:
> You can not always catch the delete(like at the time pod deleted operator not running normal), you need to check time from time.

No need to catch the event. We just check the current tikv pod. While pod is deleted and operator is not running, delete event will miss but evict-leader scheduler can be removed after operator is restarted.

> another possible issue is we may also delete evict-leader scheduler not added by us.

It seems a common issue for both of two ways?

Reply:
> No need to catch the event. We just check the current tikv pod. While pod is deleted and operator is not running, delete event will miss but evict-leader scheduler can be removed after operator is restarted.
--

yes, you need to check all clusters as mentioned before. (if not annotation but having an evict-leader scheduler, we delete the scheduler)


> It seems a common issue for both of two ways?

no, we have the annotation at tc, we will not delete the scheduler added by the user.

Reply:
Fine, I understand your concern. Let me rethink it because I prefer to act it in pod level.